routers/api/packages/npm/npm.go

```diff
@@ -0,0 +1,309 @@
+// Copyright 2021 The Gitea Authors. All rights reserved.
+// Use of this source code is governed by a MIT-style
+// license that can be found in the LICENSE file.
+
+package npm
+
+import (
+	"bytes"
+	"errors"
+	"fmt"
+	"io"
+	"net/http"
+	"net/url"
+	"strings"
+
+	"code.gitea.io/gitea/models/db"
+	packages_model "code.gitea.io/gitea/models/packages"
+	"code.gitea.io/gitea/modules/context"
+	packages_module "code.gitea.io/gitea/modules/packages"
+	npm_module "code.gitea.io/gitea/modules/packages/npm"
+	"code.gitea.io/gitea/modules/setting"
+	"code.gitea.io/gitea/routers/api/packages/helper"
+	packages_service "code.gitea.io/gitea/services/packages"
+
+	"github.com/hashicorp/go-version"
+)
+
+// errInvalidTagName indicates an invalid tag name
+var errInvalidTagName = errors.New("The tag name is invalid")
+
+func apiError(ctx *context.Context, status int, obj interface{}) {
+	helper.LogAndProcessError(ctx, status, obj, func(message string) {
+		ctx.JSON(status, map[string]string{
+			"error": message,
+		})
+	})
+}
+
+// packageNameFromParams gets the package name from the url parameters
+// Variations: /name/, /@scope/name/, /@scope%2Fname/
+func packageNameFromParams(ctx *context.Context) (string, error) {
+	scope := ctx.Params("scope")
+	id := ctx.Params("id")
+	if scope != "" {
+		return fmt.Sprintf("@%s/%s", scope, id), nil
+	}
+	return url.QueryUnescape(id)
```

Comment:
`QueryUnescape` seems incorrect. IIRC `Params` should already do the PathUnescape.

And the `fmt.Sprintf("@%s/%s", scope, id)` above doesn't do any escaping

Reply:
> `QueryUnescape` seems incorrect. IIRC `Params` should already do the PathUnescape.

Tested it again and it's not needed.

> And the `fmt.Sprintf("@%s/%s", scope, id)` above doesn't do any escaping

And it should not. This method gets called from multiple endpoints: `/{id}/` and `/@{scope}/{id}/`
If it's the first route the name is only the `id`, if it's the second route it's `@scope/id`.